selfdrive/ui/qt/widgets/cameraview.cc

```diff
@@ -237,40 +220,55 @@ void CameraViewWidget::paintGL() {
   glBindVertexArray(0);
 }
 
-void CameraViewWidget::updateFrame() {
-  if (!isVisible()) {
-    return;
+void CameraViewWidget::vipcConnected(VisionIpcClient *vipc_client) {
+  makeCurrent();
+  for (int i = 0; i < vipc_client->num_buffers; i++) {
+    texture[i].reset(new EGLImageTexture(&vipc_client->buffers[i]));
+
+    glBindTexture(GL_TEXTURE_2D, texture[i]->frame_tex);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+
+    // BGR
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_R, GL_BLUE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_G, GL_GREEN);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_B, GL_RED);
+    assert(glGetError() == GL_NO_ERROR);
   }
+  latest_frame = nullptr;
+  stream_width = vipc_client->buffers[0].width;
+  stream_height = vipc_client->buffers[0].height;
+  updateFrameMat(width(), height());
+}
+
+void CameraViewWidget::vipcFrameReceived(VisionBuf *buf) {
+  latest_frame = buf;
+  update();
+}
+
+void CameraViewWidget::vipcThread() {
+  VisionStreamType cur_stream_type = stream_type;
+  std::unique_ptr<VisionIpcClient> vipc_client_;
+
+  while (!QThread::currentThread()->isInterruptionRequested()) {
+    if (!vipc_client_ || cur_stream_type != stream_type) {
+      cur_stream_type = stream_type;
+      vipc_client_.reset(new VisionIpcClient("camerad", cur_stream_type, true));
+    }
 
-  if (!vipc_client->connected) {
-    makeCurrent();
-    if (vipc_client->connect(false)) {
-      // init vision
-      for (int i = 0; i < vipc_client->num_buffers; i++) {
-        texture[i].reset(new EGLImageTexture(&vipc_client->buffers[i]));
-
-        glBindTexture(GL_TEXTURE_2D, texture[i]->frame_tex);
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-
-        // BGR
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_R, GL_BLUE);
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_G, GL_GREEN);
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_B, GL_RED);
-        assert(glGetError() == GL_NO_ERROR);
+    if (!vipc_client_->connected) {
+      if (!vipc_client_->connect(false)) {
+        QThread::msleep(100);
+        continue;
       }
-      latest_frame = nullptr;
-      resizeGL(width(), height());
+      emit vipcThreadConnected(vipc_client_.get());
     }
-  }
 
-  VisionBuf *buf = nullptr;
-  if (vipc_client->connected) {
-    buf = vipc_client->recv(nullptr, 0);
-    if (buf != nullptr) {
-      latest_frame = buf;
-      update();
-      emit frameUpdated();
+    VisionBuf *buf = vipc_client_->recv(nullptr, 1000);
+    if (buf) {
+      emit vipcThreadFrameReceived(buf);
+    } else {
+      LOGE("visionIPC receive timeout");
```

Comment:
let's remove this log. we already have the "slow frame" print in the onroad widget